src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs

```diff
@@ -0,0 +1,189 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+using System.Collections.Immutable;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class StatelessFuncAnalyzer : DiagnosticAnalyzer {
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.StatelessFuncIsnt
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private void RegisterAnalysis( CompilationStartAnalysisContext context ) {
+
+			Compilation compilation = context.Compilation;
+
+			context.RegisterSyntaxNodeAction(
+				ctx => {
+					if( ctx.Node is ObjectCreationExpressionSyntax expr ) {
```

Comment:
more nits:

This case will always succeed, feel free to change it to `AnalyzerInvocation( ctx, ctx.Node as ObjectCreationExpression )` (or you could do the cast in `AnalyzeInvocation` and just pass `AnalyzeInvocation` directly here.

Consider renaming it to `AnalyzeObjectCreationExpression` since they're different things.